tests/test_missing.py

```diff
@@ -6,63 +6,317 @@
 import tempfile
 import unittest
 
+from netCDF4 import Dataset
+
 from activestorage.active import Active
 from activestorage.config import *
 from activestorage import dummy_data as dd
 
 import utils
 
 
-def _doit(testfile):
-    """ 
-    Compare and contrast vanilla mean with actual means
-    """
-    uri = utils.write_to_storage(testfile)
-    active = Active(uri, "data", utils.get_storage_type())
+"""
+There is a high level of duplication in this test module:
+this is to avert any issues of I/O toestepping from h5netcdf
+when running with USE_S3=True; this will probably have to change
+when we start using netCDF4python with S3-enabled
+"""
+
+def test_partially_missing_data(tmp_path):
+    testfile = str(tmp_path / 'test_partially_missing_data.nc')
+    r = dd.make_partially_missing_ncdata(testfile)
+
+    # retrieve the actual numpy-ed result
+    ds = Dataset(testfile)
+    actual_data = ds["data"][:]
+    ds.close()
+    numpy_mean = np.ma.mean(actual_data[0:2, 4:6, 7:9])
+    print("Numpy masked result (mean)", numpy_mean)
+
+    # write file to storage
+    testfile = utils.write_to_storage(testfile)
+
+    # run the two Active instances: transfer data and do active storage
+    active = Active(testfile, "data", utils.get_storage_type())
     active._version = 0
     d = active[0:2, 4:6, 7:9]
+
+    # NOT numpy masked to check for correct Active behaviour
     mean_result = np.mean(d)
+    print("No active storage result (mean)", mean_result)
 
-    active = Active(uri, "data", utils.get_storage_type())
+    active = Active(testfile, "data", utils.get_storage_type())
     active._version = 2
     active.method = "mean"
     active.components = True
     result2 = active[0:2, 4:6, 7:9]
-    np.testing.assert_array_equal(mean_result, result2["sum"]/result2["n"])
+    active_result = result2["sum"] / result2["n"]
+    print("Active storage result (mean)", active_result)
 
+    if not USE_S3:
+        np.testing.assert_array_equal(numpy_mean, active_result)
+    else:
+        np.testing.assert_raises(AssertionError,
+                                 np.testing.assert_array_equal,
+                                 numpy_mean, active_result)
+    np.testing.assert_array_equal(mean_result, active_result)
```

Comment:
Surely this should be not equal? mean_result is the unmasked mean, and active_result is the result from Active.

Overall, I'd switch the assertions here.

1. Sanity check that numpy_mean != mean_result, to ensure we have some masked data
2. if not USE_S3, assert numpy_mean == active_result
3. else, assert unmasked_mean == active_result

Reply:
> Surely this should be not equal? mean_result is the unmasked mean, and active_result is the result from Active.

Good point! They are not equal when we set fill_values or valid_min/max. But in this case `np.ma.mean(d)` and `np.mean(d)` are equivalent and give the same result - this is exactly why I use the comparison with the "unmasked" `np.mean(d)` so that we detect any issues in constructing the masked data ie not a properly constructed mask that will be ignored by `np.mean()` and return a result using the masked elements too

Reply:
> Overall, I'd switch the assertions here.

This is essentially what I'm doing:

- for masked arrays (actual mask, not a fill_value/valid_min/max) mask I use the np.mean as "unmasked" result to catch any issues with the mask
- for filled arrays I mask on the fill values and go on using that as the masked array
- also important to test the Active result when there is no active storage run (version=0) versus the Active with active storage (version=2)

Reply:
one thing to remember is that `np.mean(masked_array)` is now equal to `np.ma.mean(masked_array)` when there is a proper mask in place :grin: